Dockerfile

```diff
@@ -1,17 +1,34 @@
-FROM node:12-alpine
-
+FROM node:12-alpine as base
 WORKDIR /app
-
 COPY package*.json ./
+ENV PORT=8080
 
-RUN npm install
-
+FROM base AS dev
+RUN npm ci
 COPY . .
+CMD ["npm", "run", "dev"]
 
+FROM dev AS test
+COPY . .
+CMD ["npm", "run", "test"]
+
+# EXPOSE 8080
+FROM base AS builder
+RUN npm ci
+COPY src src
+COPY webpack* .
+COPY .eslintrc.json .
+COPY .babelrc .
+COPY tsconfig.json .
 RUN npm run build
 
-ENV PORT=8080
-
-EXPOSE 8080
-
+FROM base AS prod
+# this still installs all the non-dev dependencies; removing react etc would be better
+RUN npm ci --production
+COPY --from=builder /app/build ./build
+COPY app.js app.js
 CMD ["node", "app.js"]
+
+FROM nginx:mainline-alpine AS nginx
+COPY --from=builder /app/build /usr/share/nginx/html
```

Comment:
This is the magic of multistage dockerfiles: we use an entirely separate image(nginx) to host the web server. This image literally doesnt even have node, and it has none of the things in this repo actually except for default.conf (and the webpack output). We use the --from=builder option to let Docker know that this image depends on the builder image, so it builds the builder before this one.

Reply:
Multistage dockerfiles also allow for parallelism; it finds layers that have no dependencies and runs them simultaneously, blocking when there's a dependency such as copy --from